### PR TITLE
IBP-4006 multi app single server impl

### PR DESCRIPTION
**Pull request by grg021**
added config files for multiple bmsapp instance
added setup script
updated the readme documentation

#IBP-4006

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,6 +2,18 @@
 * docker and docker-compose is installed
 * IPV4 forwarding is enabled - net.ipv4.ip_forward=1
 * a [dockerhub](https://hub.docker.com/) account with access to the bmsapp repo. For external dev teams access requests can be performed via our [HelpDesk](https://ibplatform.atlassian.net/servicedesk/customer/portal/4/group/30/create/60)
+
+## Configurations
+* **.env.example** - this is where we setup environment variables utilized by the docker compose configuration. In order to use it it must be renamed to .env
+* **config** - this contains the configuration files for the BMSScripts inside bmsapp.
```
Comment (darla-leafnode) [inline README.md]: Change to `DbScripts` (instead of BMSScripts)

File: README.md [hunk 2]
```diff
@@ -2,6 +2,18 @@
...
+* **mysql/conf** - this contains mysql configuration that is bind mounted to the mysql container on launch.
+* **nginx** - this contains nginx configuration that is bind mounted to the nginx-proxy container on launch.
+* **docker-compose.yml** - this default configuration is a basic setup of bmsapp with its own database that can be ported to 8080 or 80.
+* **docker-compose-ssl.yml** - this is a setup that uses 4 containers to serve the bmsapp via HTTPS. The TLS Certificate is created and reenwed by the letsencrypt container and the requests are handled by the nginx-proxy container.
```
Comment (darla-leafnode) [inline README.md]: 1. correct typo: **renewed**
2. Let's Encrypt

File: config/test.properties [hunk 3]
```diff
@@ -0,0 +1,54 @@
+########################################
+########## Workbench related properties
+########################################
+
+#Workbench hibernate configuration
```
Comment (darla-leafnode) [inline config/test.properties]: Why do we need `test.properties` here?

File: multi-docker-compose-app.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    environment:
+      - VIRTUAL_HOST=${BMS_SSL_HOST:-bmsdock.leafnode.io}
+      - VIRTUAL_PORT=18080
+      - LETSENCRYPT_HOST=${BMS_SSL_HOST:-bmsdock.leafnode.io}
+      - LETSENCRYPT_EMAIL=${BMS_SSL_EMAIL:-leafnodedev@leafnode.io}
+      - BMS_SCHEME=${BMS_SCHEME:-https}
+      - BMS_PORT=${BMS_PORT:-443}
+      - BMS_CROPS=${BMS_CROPS:?err}
+      - BMS_DB_HOST=${BMS_DB_HOST:?err}
+      - BMS_DB_PASSWORD=${BMS_DB_PASS:?err}
```
Comment (darla-leafnode) [inline multi-docker-compose-app.yml:19]: Should this be: **BMS_DB_PASS**=${BMS_DB_PASS:?err}

File: docker-compose-ssl.yml [hunk 5]
```diff
@@ -3,6 +3,7 @@ services:
   nginx-proxy:
     image: jwilder/nginx-proxy
     container_name: nginx-proxy
+    restart: always
```
Comment (darla-leafnode) [inline docker-compose-ssl.yml:6]: Why do we set to restart = always?

File: docker-compose-ssl.yml [hunk 6]
```diff
@@ -3,6 +3,7 @@ services:
   nginx-proxy:
```
Comment (darla-leafnode) [inline docker-compose-ssl.yml:3]: For bmsapp container, add the BMS_PASSWORD environment property.
Also do the same in docker-compose.yml file

File: .env.sample [hunk 7]
```diff
@@ -1,5 +1,6 @@
 BMS_IMG_URI=ibpbms/bmsapp
 BMS_RELEASE=15.0
```
Comment (darla-leafnode) [inline .env.sample]: Change BMS_RELEASE to 16.0

File: setup.sh [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    if [ "$RUN" != "y" ]; then
+            die "Setup discontinued."
+    fi
+
+    mv docker-compose.yml docker-compose.yml.bak
+    cp  multi-docker-compose-app.yml docker-compose.yml
+    cp .env.sample .env
+    cp config/application.properties.orig config/application.properties
+
+    sed -i -e "s/BMS_RELEASE=15.0/BMS_RELEASE=$BMS_VERSION/g" .env
```
Comment (darla-leafnode) [inline setup.sh]: Change BMS_RELEASE to 16.0

**Review by darla-leafnode — CHANGES_REQUESTED** [review]
Please see my comments/questions @grg021

File: docker-compose-ssl.yml (hunk 5 above)
Comment (grg021) [reply docker-compose-ssl.yml:6]: For nginx-proxy we need it to always be running. When the host restarts for example we need it also to start automatically.

File: README.md (hunk 1 above)
Comment (grg021) [reply README.md]: ok

File: README.md (hunk 2 above)
Comment (grg021) [reply README.md]: ok thanks

File: docker-compose-ssl.yml (hunk 6 above)
Comment (grg021) [reply docker-compose-ssl.yml:3]: done

File: .env.sample (hunk 7 above)
Comment (grg021) [reply .env.sample]: done

File: setup.sh (hunk 8 above)
Comment (grg021) [reply setup.sh]: done

File: config/test.properties (hunk 3 above)
Comment (grg021) [reply config/test.properties]: removed extra file

File: docker-compose-ssl.yml (hunk 5 above)
Comment (grg021) [reply docker-compose-ssl.yml:6]: This restarts the nginx-proxy upon start of the server or restart. This is set for all the containers.

File: docker-compose.yml [hunk 9]
```diff
@@ -3,18 +3,21 @@ services:
...
     depends_on:
       - bmsmysql
-    networks:
-      - bmsnet
     image: ${BMS_IMG_URI:?err}:${BMS_RELEASE:?err}
     ports:
       - "80:8080"
     restart: always
     environment:
+      - VIRTUAL_HOST=${BMS_SSL_HOST:-bmsdock.leafnode.io}
```
Comment (darla-leafnode) [inline docker-compose.yml]: I think we don't need these environment properties for non-ssl setup:
`VIRTUAL_HOST
VIRTUAL_PORT
LETSENCRYPT_HOST
LETSENCRYPT_EMAIL
BMS_SCHEME
BMS_PORT`
Right?

File: docker-compose.yml [hunk 10]
```diff
@@ -26,8 +29,6 @@ services:
     image: mysql:5.6
     ports:
        - "3307:3306"
-    networks:
```
Comment (darla-leafnode) [inline docker-compose.yml:29]: Why did we remove the `networks `config?

File: docker-compose.yml [hunk 11]
```diff
@@ -3,18 +3,21 @@ services:
...
+      - VIRTUAL_PORT=8080
+      - LETSENCRYPT_HOST=${BMS_SSL_HOST:-bmsdock.leafnode.io}
+      - LETSENCRYPT_EMAIL=${BMS_SSL_EMAIL:-leafnodedev@leafnode.io}
+      - BMS_SCHEME=${BMS_SCHEME:-https}
+      - BMS_PORT=${BMS_PORT:-443}
       - BMS_CROPS=${BMS_CROPS:?err}
+      - BMS_DB_PASSWORD=${BMS_DB_PASS:?err}
     volumes:
       - mysqldata:/var/lib/mysql
-    links:
```
Comment (darla-leafnode) [inline docker-compose.yml:16]: Why did you remove the `links `config?

**Review by darla-leafnode — CHANGES_REQUESTED** [review]
Please check my further comments @grg021

File: docker-compose.yml (hunk 11 above)
Comment (grg021) [reply docker-compose.yml:16]: the links flag is a legacy feature and not required anymore. were already linking them through the network https://docs.docker.com/compose/compose-file/#links

File: docker-compose.yml (hunk 10 above)
Comment (grg021) [reply docker-compose.yml:29]: removed this in the multi container setup but for this config I reverted it back since were not using the default network

**Comment by grg021** [comment]
Please see updated config file @darla-leafnode

**Review by darla-leafnode — APPROVED** [review]
